OutOfSchool/OutOfSchool.WebApi/Controllers/V1/WorkshopController.cs

```diff
@@ -361,7 +361,14 @@ public async Task<IActionResult> Update([FromBody] WorkshopBaseDto dto)
             return StatusCode(403, "Forbidden to update workshops, which are not related to you");
         }
 
-        if (!await combinedWorkshopService.IsAvailableSeatsValid(dto.AvailableSeats, dto.Id))
+        var workshop = await combinedWorkshopService.GetById(dto.Id, true).ConfigureAwait(false);
+
+        if (workshop is null)
+        {
+            return BadRequest("Workshop does not exist.");
+        }
+
+        if (!combinedWorkshopService.IsAvailableSeatsValid(dto.AvailableSeats, workshop))
```

Comment:
(Design smell) Please consider moving such calls closer to the implementation, because now we have some sort of "jealous method" smell: controller action, that doesn't trust `combinedWorkshopService` and instead is calling each method (`.GetById()`, `.IsAvailableSeatsValid()`, `.Update()`) by itself.
Ideally, all logic related to data validation that is dependent on db data should be done in service.

Reply:
I made the correction if I understood you correctly. Thank you!

Reply:
@DKymachynskyi I was talking about all three methods  (`.GetById()`, `.IsAvailableSeatsValid()`, `.Update()`) , not only first two. 

I'm not insisting on next changes, it's up to you to decide do it or not, I'm just sharing my point.

So currently there is such code in controller
```
        var result = await combinedWorkshopService.IsAvailableSeatsValidForWorkshop(
            dto.AvailableSeats, dto.Id).ConfigureAwait(false);

        if (result is null)
        {
            return BadRequest("The workshop does not exist.");
        }

        if (result == false)
        {
            return BadRequest("The number of available seats must be equal or greater than the number of taken seats");
        }

        try
        {
            var updatingResult = await combinedWorkshopService.Update(dto).ConfigureAwait(false);
            return Ok(CreateUpdateResponse(updatingResult));
        }
        catch (ArgumentException e)
        {
            return BadRequest(e.Message);
        } 
```
my suggestion is to have this code reduced to simple
```
        var result = await combinedWorkshopService.Update(dto).ConfigureAwait(false);

        if (!result.Succeded)
        {
            return BadRequest(result.OperationResult.ToString());
        }

        return Ok(CreateUpdateResponse(result).Value);
```
where all previous checks are moved inside `combinedWorkshopService.Update()` that returns `Task<Result<WorkshopBaseDto>>` and all checks are done inside and errors are stored as `Result.OperationResult.Errors`.

(Minor note between lines - design of `Result` and `OperationResult` types is pretty hectic, but it works, at least)

Thus controller has no logic except dto mapping (that `CreateUpdateResponse()`) and you have better test locality for logic in your code.